Generate marker gene references for CellAssign

allyhawkins opened the pull request:
Closes #384 

This PR builds on the additions from #388 and adds in a process for creating the marker gene x cell type matrices need for `CellAssign`. 

- I created a script that takes as input a marker gene list, a list of organs, and a gtf file and outputs a binary gene x cell type matrix with ensembl ID's. I used the same method as our data integration process to create the matrix. First, I filtered the larger marker gene table by organs and then converted the resulting gene list into a binary matrix.
- I did add options to the script for the columns that contain the cell type and the gene id. In case we ever want to use a different marker table as the reference we could still use this script but change those arguments. The only thing that I don't make optional is the `organs`, because we are specifying that in the metadata table. 
- I wanted the output to contain the genes as ensembl IDs and not as human gene symbols, which is how PanglaoDB shows them. Previously, we were grabbing the ensembl IDs from the rownames of the SCE object we are annotating. Here, we want to create a single reference that can be used on all SCE objects, so I imported the `gtf` file that we use to build our reference and grab the ids from there. 
- I tested this and the new marker gene matrices are in `s3://scpca-references/celltype/cellassign_references`

sjspielman commented on bin/generate_cellassign_refs.R:
```diff
@@ -0,0 +1,120 @@
+# grab marker genes for specified organs from marker gene reference
+organ_gene_df <- marker_gene_df |>
+  dplyr::filter(organ %in% organs) |>
+  dplyr::select({{cell_type_column}}, {{gene_id_column}})
```
Oooh I've never used `{{}}` like this before but I like it! I would have done something like - 
```
dplyr::select(all_of(c(cell_type_column, gene_id_column)))
```

....but what you have done is so much cleaner :)

sjspielman commented on bin/generate_cellassign_refs.R:
```diff
@@ -0,0 +1,120 @@
+    "ensembl_id" = "gene_id",
+    "gene_symbol" = "gene_name"
+  ) |>
+  dplyr::filter(.data$gene_symbol != "NA") |>
+  dplyr::distinct() |>
+  ## in case there are any duplicate gene_ids (there shouldn't be!)
+  dplyr::group_by(.data$ensembl_id) |>
+  dplyr::summarise(gene_symbol = paste(.data$gene_symbol, collapse = ";"))
```
we don't need `.data$` right? (also i snuck a `z` in there 🇺🇸 )


```suggestion
  dplyr::filter(gene_symbol != "NA") |>
  dplyr::distinct() |>
  ## in case there are any duplicate gene_ids (there shouldn't be!)
  dplyr::group_by(ensembl_id) |>
  dplyr::summarize(gene_symbol = paste(gene_symbol, collapse = ";"))
```

allyhawkins replied:
Whoops... I took a lot of this directly from the function that adds gene symbols in `scpcaTools`.

sjspielman commented on bin/generate_cellassign_refs.R:
```diff
@@ -0,0 +1,120 @@
+    help = "File containing a list of marker genes for all cell types and all organs. 
+      Must contain the `organ` column, `--cell_type_column` and `--gene_id_column`. "
+  ),
+  make_option(
+    opt_str = c("--cell_type_column"),
+    type = "character",
+    default = "cell type"
+  ),
+  make_option(
+    opt_str = c("--gene_id_column"),
+    type = "character",
+    default = "official gene symbol"
+  ),
```
Can we add `help` here?

sjspielman commented on build-celltype-ref.nf:
```diff
@@ -54,6 +55,32 @@ process train_singler_models {
+  publishDir "${params.celltype_ref_dir}/cellassign_references"
+  label 'mem_8'
+  input:
+    tuple val(ref_name), val(ref_database), val(organs)
```
```suggestion
    tuple val(ref_name), val(ref_source), val(organs)
```

sjspielman replied:
I guess this reference _is_ a database, but suggesting for overall consistency

sjspielman commented on build-celltype-ref.nf:
```diff
@@ -54,6 +55,32 @@ process train_singler_models {
+  output:
+    path ref_file
+  script:
+    ref_file="${ref_database}-${ref_name}.tsv"
```
```suggestion
    ref_file="${ref_source}-${ref_name}.tsv"
```

sjspielman commented on build-celltype-ref.nf:
```diff
@@ -54,6 +55,32 @@ process train_singler_models {
+      --ref_mtx_file ${ref_file}
+    """
+  stub:
+    ref_file="${ref_database}-${ref_name}.tsv"
```
```suggestion
    ref_file="${ref_source}-${ref_name}.tsv"
```

sjspielman commented:
Overall looks good to me, just a couple comments to solidify it a bit!

allyhawkins commented:
Thanks for the review @sjspielman! I incorporated your suggestions and then also added in the help messages that were missing. This is ready for another look.

sjspielman reviewed (approved):
🚀
